### PR TITLE
AA-722: Progress Tab groundwork

### PR DESCRIPTION
**Pull request by ciduarte**
## Components covered in this PR:
- [x] Grade Summary
- [x] Detailed Grades
- [x] Related Links

There are placeholders in this PR for Course Grade, Course Completion, and Certificate Status that have been cleaned up from the hackathon. Development of those components is covered in future tickets.

[Link to mocks](https://www.figma.com/file/KFz3P2Is0nmIBzy4nf4gJy/Progress-tab?node-id=318%3A2155)

### Grade Summary and Detailed Grades, Desktop and Mobile
AA-722_GradeSummary_DetailedGrades_Desktop
<div>

**Non-passing and Passing**

AA-722_GradeSummary_DetailedGrades_Mobile
AA-722_GradeSummary_Passing
</div>

### Related Links, Desktop and Mobile
<div>
AA-722_RelatedLinks_Desktop
AA-722_RelatedLinks_Mobile
</div>

### CODE REVIEW

File: src/course-home/progress-tab/certificate-status/CertificateStatus.jsx [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import React from 'react';
+
+function CertificateStatus() {
+  return (
+    <section className="text-dark-700 rounded shadow-sm my-4 p-4">
+      {/* TODO: AA-719 */}
```
Comment (ciduarte) [inline src/course-home/progress-tab/certificate-status/CertificateStatus.jsx:6]: Placeholder

File: src/course-home/progress-tab/course-completion/CourseCompletion.jsx [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import React from 'react';
+import { useSelector } from 'react-redux';
+
+import { useModel } from '../../../generic/model-store';
+
+function CourseCompletion() {
+  // TODO: AA-720
```
Comment (ciduarte) [inline src/course-home/progress-tab/course-completion/CourseCompletion.jsx:6]: Placeholder, feel free to ignore

File: src/course-home/progress-tab/grades/course-grade/CourseGrade.jsx [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import React from 'react';
+
+function CourseGrade() {
+  return (
+    <section className="text-dark-700 my-4 rounded shadow-sm p-4">
+      {/* TODO: AA-721 */}
```
Comment (ciduarte) [inline src/course-home/progress-tab/grades/course-grade/CourseGrade.jsx:6]: Placeholder

File: src/course-home/progress-tab/grades/grade-summary/GradeSummary.jsx [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  });
+
+  sectionScores.forEach((chapter) => {
+    chapter.subsections.forEach((subsection) => {
+      if (subsection.hasGradedAssignment) {
+        gradeByAssignmentType[subsection.assignmentType].numPointsEarned += subsection.numPointsEarned;
+        gradeByAssignmentType[subsection.assignmentType].numPointsPossible += subsection.numPointsPossible;
+      }
+    });
+  });
```
Comment (ciduarte) [inline src/course-home/progress-tab/grades/grade-summary/GradeSummary.jsx:37]: Trying to link assignment types to grades, doing this felt questionable. Is this fine or gross?

File: src/course-home/progress-tab/course-completion/CourseCompletion.jsx [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  const {
+    completionSummary: {
+      completeCount,
+      incompleteCount,
+      lockedCount,
+    },
+  } = useModel('progress', courseId);
+
+  const total = completeCount + incompleteCount + lockedCount;
+  const completePercentage = ((completeCount / total) * 100).toFixed(0);
```
Comment (mikix) [inline src/course-home/progress-tab/course-completion/CourseCompletion.jsx:20]: oooh I didn't know about `toFixed` and I will probably forget it again, but that's nice.

File: src/course-home/progress-tab/grades/grade-summary/DroppableAssignmentFootnote.jsx [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+function DroppableAssignmentFootnote({ footnotes, marker, intl }) {
+  return (
+    <>
+      <span id="grade-summary-footnote-label" className="sr-only">{intl.formatMessage(messages.footnotesTitle)}</span>
+      <ul className="list-unstyled mt-2">
+        {footnotes.map((footnote, index) => (
+          <li id={`${footnote.id}-footnote`} key={footnote.id} className="x-small mt-1">
+            <FormattedMessage
+              id="progress.footnotes.droppableAssignments"
+              defaultMessage="{footnoteMarker} The lowest {numDroppable} {assignmentType} {score} will be dropped."
```
Comment (mikix) [inline src/course-home/progress-tab/grades/grade-summary/DroppableAssignmentFootnote.jsx]: This whole footnote thing is complicated, but I like it - I mean, it's necessary complication and probably looks/works nicely.

File: src/course-home/progress-tab/grades/grade-summary/DroppableAssignmentFootnote.jsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+              values={{
+                footnoteMarker: marker.repeat(index + 1),
+                numDroppable: footnote.numDroppable,
+                assignmentType: footnote.assignmentType,
+                score: footnote.numDroppable > 1 ? 'scores' : 'score',
```
Comment (mikix) [inline src/course-home/progress-tab/grades/grade-summary/DroppableAssignmentFootnote.jsx]: That's not going to work for many languages. Lots have weird pluralization rules. There might be a FormattedMessage variant for handling numbers of things / plurals?

File: src/course-home/progress-tab/grades/grade-summary/GradeSummary.jsx (hunk 4 above)
Comment (mikix) [reply src/course-home/progress-tab/grades/grade-summary/GradeSummary.jsx:37]: Yours might be the clearest way? Another way could be some `reduce` calls, but you'd have to nest them and do it twice, once for earned and once for possible, right? Your way is fine.

File: src/course-home/progress-tab/grades/grade-summary/GradeSummaryTable.jsx [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+  const footnotes = [];
+  const asterisk = '*';
+
+  const calculateWeightedGrade = (numPointsEarned, numPointsPossible, assignmentWeight) => (
+    numPointsPossible > 0 ? ((numPointsEarned * assignmentWeight * 100) / numPointsPossible).toFixed(0) : 0
+  );
+
+  const getFootnoteId = (assignment) => (
+    `${assignment.shortLabel ? assignment.shortLabel : (assignment.type).replace(/\s+/g, '-')}`
```
Comment (mikix) [inline src/course-home/progress-tab/grades/grade-summary/GradeSummaryTable.jsx]: I'm not entirely sure where footnote ids get used - but spaces are the only invalid character? And `shortLabel` is guaranteed not to have them?

File: src/course-home/progress-tab/grades/grade-summary/GradeSummaryTable.jsx [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+    if (assignment.numDroppable > 0) {
+      footnoteId = getFootnoteId(assignment);
+      footnotes.push({
+        id: footnoteId,
+        numDroppable: assignment.numDroppable,
+        assignmentType: assignment.type,
+      });
+
+      footnoteMarker = asterisk.repeat(footnotes.length);
```
Comment (mikix) [inline src/course-home/progress-tab/grades/grade-summary/GradeSummaryTable.jsx]: I'm not sure I love this? What about numbers instead?

Or if you want to use symbols, cycle through a few before doubling. Wikipedia says the standard order is *, †, ‡, §, ‖, ¶ - and then start repeating. (Though I do like this quote in its own footnote for that list: `But beyond the ... double dagger, this order is not familiar to most readers, and never was.` Still though, adding some variety to the symbols is nice.)

https://en.wikipedia.org/wiki/Note_(typography)

File: src/course-home/progress-tab/grades/grade-summary/AssignmentTypeCell.jsx [hunk 10]
```diff
@@ -0,0 +1,28 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+function AssignmentTypeCell({ assignmentType, footnoteMarker, footnoteId }) {
+  return (
+    <div>
+      {assignmentType}
+      {footnoteId && footnoteMarker && (
+        <a id={`${footnoteId}-ref`} className="text-dark-700" href={`#${footnoteId}-footnote`} aria-describedby="grade-summary-footnote-label">
+          {footnoteMarker}
```
Comment (mikix) [inline src/course-home/progress-tab/grades/grade-summary/AssignmentTypeCell.jsx]: Do you want a `<sup>` element here too?

**Review by mikix — APPROVED** [review]
Did not run this locally, but the screenshots look great.

It's obviously a WIP with all the placeholders, but I apparently have lots of thoughts about footnotes. :P

File: src/course-home/progress-tab/grades/grade-summary/GradeSummaryTable.jsx (hunk 9 above)
Comment (ciduarte) [reply src/course-home/progress-tab/grades/grade-summary/GradeSummaryTable.jsx]: The asterisks were a Product/Design decision, but I can ask them what they prefer here

File: src/course-home/progress-tab/grades/grade-summary/GradeSummaryTable.jsx (hunk 8 above)
Comment (ciduarte) [reply src/course-home/progress-tab/grades/grade-summary/GradeSummaryTable.jsx]: This eventually ends up being `<div id={footnoteId}></div>`, and the guidance there is to not include whitespace in an element's id. But looking into it, there's further guidance to remove non-ascii characters and all symbols that aren't '-', '_', or '.' as well, so I'll add that in. `shortLabel` seems to usually be an abbreviation, but also no guarantees there. Thanks for pointing that out.

File: src/course-home/progress-tab/grades/grade-summary/DroppableAssignmentFootnote.jsx (hunk 7 above)
Comment (ciduarte) [reply src/course-home/progress-tab/grades/grade-summary/DroppableAssignmentFootnote.jsx]: There was 😁

File: src/course-home/progress-tab/grades/grade-summary/GradeSummaryTable.jsx (hunk 9 above)
Comment (ciduarte) [reply src/course-home/progress-tab/grades/grade-summary/GradeSummaryTable.jsx]: They didn't have any strong opinions here, so I can go with numbers.

**Pull request merged**